src/Adapter/Product/Stock/Repository/StockMovementRepository.php

```diff
@@ -56,25 +54,133 @@ public function __construct(
+     * The range filter is then applied as 2nd selection of single histories,
+     * whereas the remaining stock movements are grouped into range histories.
+     *
+     * @return array<int, array<string, mixed>>
+     */
+    public function getLastStockMovementHistories(
+        StockMovementFilter $mainFilter,
+        StockMovementFilter $rangeFilter,
+        int $offset = 0,
+        int $limit = self::DEFAULT_LIMIT
+    ): array {
+        $queryBuilder = $this
+            ->createFilterQueryBuilder($mainFilter, $offset, $limit)
+            ->select(
+                'MIN(sm.id_stock_mvt) id_stock_mvt_min',
+                'COUNT(id_stock_mvt) id_stock_mvt_count',
+                'GROUP_CONCAT(id_stock_mvt) id_stock_mvt_list',
+                'GROUP_CONCAT(id_stock) id_stock_list',
+                'GROUP_CONCAT(id_stock_mvt_reason) id_stock_mvt_reason_list',
+                'GROUP_CONCAT(id_order) id_order_list',
+                'GROUP_CONCAT(id_employee) id_employee_list',
+                'MIN(employee_firstname) employee_firstname',
+                'MIN(employee_lastname) employee_lastname',
+                'SUM(sm.sign * sm.physical_quantity) delta_quantity',
+                'MIN(sm.date_add) date_add_min',
+                'MAX(sm.date_add) date_add_max'
+            )
+            ->orderBy('id_stock_mvt_min', 'DESC')
+        ;
+        $this->updateQueryBuilderWithGroupByRange(
+            $queryBuilder,
+            $this->createFilterQueryBuilder($rangeFilter)
+        );
 
-        return $qb->execute()->fetchAllAssociative();
+        return $queryBuilder->execute()->fetchAllAssociative();
+    }
+
+    /**
+     * Updates a query builder with a filter to group irrelevant rows in the same ranges
+     *
+     * @param QueryBuilder $queryBuilder Query builder to update
+     * @param QueryBuilder $rangeQueryBuilder Query builder used as a range filter
+     * @param string $pkColumn Primary key column name
+     * @param string $rangeColumn Column name used to detect ranges
+     * @param string $groupColumn Column name used to group rows into ranges
+     */
+    protected function updateQueryBuilderWithGroupByRange(
+        QueryBuilder $queryBuilder,
+        QueryBuilder $rangeQueryBuilder,
+        string $pkColumn = 'sm.id_stock_mvt',
+        string $rangeColumn = 'range_id',
+        string $groupColumn = 'range_name'
+    ): void {
+        $rangeCondition = (string) $rangeQueryBuilder->getQueryPart('where');
+        $queryBuilder
+            ->addSelect(
+                implode(' ', [
+                    "MIN(@$rangeColumn := CASE",
+                    "WHEN @$rangeColumn IS NULL THEN $pkColumn",
+                    "WHEN $rangeCondition THEN $pkColumn",
+                    "ELSE @$rangeColumn",
+                    "END) $rangeColumn",
+                ]),
+                "CASE WHEN $rangeCondition THEN CONCAT('single-', $pkColumn) ELSE CONCAT('range-', @$rangeColumn) END $groupColumn"
```

Comment:
We never use double quotes, if you need to inject a variable in a string prefer using this

```suggestion
                sprintf('CASE WHEN $rangeCondition THEN CONCAT('single-', %s) ELSE CONCAT('range-', @%s) END %s', $pkColumn, $rangeColumn, $groupColumn)
```

Reply:
Same thing for the other elements in the implode

Reply:
I usually use `sprintf()` as well, but here there are too many placeholders in random order. Using `sprintf()` syntax would make this complex SQL query even more difficult to understand.
Long term maintenance should be a first priority, so I keep named variables in the double quoted string.

Reply:
Alternative solution : replace placeholders with `str_replace()`, but I suppose it would come with a performance hit.
What do you think ?